File: caribou/deployment_solver/deployment_input/components/loaders/datacenter_loader.py
from typing import Any

from caribou.common.constants import (
    PROVIDER_REGION_TABLE,
    PROVIDER_TABLE,
    SOLVER_INPUT_AVERAGE_CPU_POWER_DEFAULT,
    SOLVER_INPUT_AVERAGE_MEMORY_POWER_DEFAULT,
    SOLVER_INPUT_CFE_DEFAULT,
    SOLVER_INPUT_COMPUTE_COST_DEFAULT,
    SOLVER_INPUT_INVOCATION_COST_DEFAULT,
    SOLVER_INPUT_PUE_DEFAULT,
    SOLVER_INPUT_TRANSMISSION_COST_DEFAULT,
)
from caribou.common.models.remote_client.remote_client import RemoteClient
from caribou.deployment_solver.deployment_input.components.loader import InputLoader


class DatacenterLoader(InputLoader):
    _datacenter_data: dict[str, Any]
    _provider_data: dict[str, Any]
    _provider_table: str

    def __init__(self, client: RemoteClient) -> None:
        super().__init__(client, PROVIDER_REGION_TABLE)
        self._provider_table = PROVIDER_TABLE

    def setup(self, available_regions: set[str]) -> None:
        self._datacenter_data = self._retrieve_region_data(available_regions)

        # Get the set of providers from the available regions
        providers = set()
        for region in available_regions:
            provider, _ = region.split(":")
            providers.add(provider)

        self._provider_data = self._retrieve_provider_data(providers)  # ignore as not yet implemented

    def get_average_cpu_power(self, region_name: str) -> float:
        return self._datacenter_data.get(region_name, {}).get(
            "average_cpu_power", SOLVER_INPUT_AVERAGE_CPU_POWER_DEFAULT
        )

    def get_average_memory_power(self, region_name: str) -> float:
        return self._datacenter_data.get(region_name, {}).get(
            "average_memory_power", SOLVER_INPUT_AVERAGE_MEMORY_POWER_DEFAULT
        )

    def get_pue(self, region_name: str) -> float:
        return self._datacenter_data.get(region_name, {}).get("pue", SOLVER_INPUT_PUE_DEFAULT)

    def get_cfe(self, region_name: str) -> float:
        return self._datacenter_data.get(region_name, {}).get("cfe", SOLVER_INPUT_CFE_DEFAULT)

    def get_compute_cost(self, region_name: str, architecture: str) -> float:
        return (
            self._datacenter_data.get(region_name, {})
            .get("execution_cost", {})
            .get("compute_cost", {})
            .get(architecture, SOLVER_INPUT_COMPUTE_COST_DEFAULT)
        )

    def get_invocation_cost(self, region_name: str, architecture: str) -> float:
        return (
            self._datacenter_data.get(region_name, {})
            .get("execution_cost", {})
            .get("invocation_cost", {})
            .get(architecture, SOLVER_INPUT_INVOCATION_COST_DEFAULT)
        )

    def get_transmission_cost(self, region_name: str, intra_provider_transfer: bool) -> float:
        transfer_type = "provider_data_transfer" if intra_provider_transfer else "global_data_transfer"

        return (
            self._datacenter_data.get(region_name, {})
            .get("transmission_cost", {})
            .get(transfer_type, SOLVER_INPUT_TRANSMISSION_COST_DEFAULT)
        )

    def _retrieve_provider_data(self, available_providers: set[str]) -> dict[str, Any]:
        all_data: dict[str, Any] = {}

        for provider in available_providers:
            all_data[provider] = self._retrieve_data(self._provider_table, provider)

        return all_data

File: caribou-go/src/deployment-input/components/loaders/datacenterloader.go
package loaders

import "caribou-go/src/utils"

const (
	SolverInputAverageMemoryPowerDefault    = 100.0
	SolverInputPueDefault                   = 1.0
	SolverInputCfeDefault                   = 0.0
	SolverInputComputeCostDefault           = 100.0
	SolverInputInvocationCostDefault        = 100.0
	SolverInputTransmissionCostDefault      = 100.0
	SolverInputMaxCpuPowerDefault           = 0.0035
	SolverInputMinCpuPowerDefault           = 0.00074
	SolverInputSnsRequestCostDefault        = 0.50 / 1000000 // 0.50 USD per 1 million requests (At Ohio region)
	SolverInputDynamodbReadCostDefault      = 0.25 / 1000000 // 0.25 USD per 1 million read request unit (At Ohio region)
	SolverInputDynamodbWriteCostDefault     = 1.25 / 1000000 // 1.25 USD per 1 million write request unit (At Ohio region)
	SolverInputEcrMonthlyStorageCostDefault = 0.10           // 0.10 USD per 1 GB per month (At Ohio region)
)

type DataCenterLoader struct {
	DataCenterData map[string]interface{}
}

func NewDataCenterLoader(datacenterData map[string]interface{}) (*DataCenterLoader, error) {
	datacenterLoader := DataCenterLoader{DataCenterData: datacenterData}
	return &datacenterLoader, nil
}

func (d *DataCenterLoader) GetAverageMemoryPower(regionName string) float64 {
	return utils.Get(d.DataCenterData, SolverInputAverageMemoryPowerDefault, regionName, "average_memory_power")
}

func (d *DataCenterLoader) GetPue(regionName string) float64 {
	return utils.Get(d.DataCenterData, SolverInputPueDefault, regionName, "pue")
}

func (d *DataCenterLoader) GetCfe(regionName string) float64 {
	return utils.Get(d.DataCenterData, SolverInputCfeDefault, regionName, "cfe")
}

func (d *DataCenterLoader) GetMaxCpuPower(regionName string) float64 {
	return utils.Get(d.DataCenterData, SolverInputMaxCpuPowerDefault, regionName, "max_cpu_power_kWh")
}

func (d *DataCenterLoader) GetMinCpuPower(regionName string) float64 {
	return utils.Get(d.DataCenterData, SolverInputMinCpuPowerDefault, regionName, "min_cpu_power_kWh")
}

func (d *DataCenterLoader) GetSNSRequestCost(regionName string) float64 {
	return utils.Get(d.DataCenterData, SolverInputSnsRequestCostDefault, regionName, "sns_cost", "sns_cost")
}

func (d *DataCenterLoader) GetDynamoDBReadWriteCost(regionName string) (float64, float64) {
	dynamoDbCosts := utils.Get(d.DataCenterData, map[string]interface{}{}, regionName, "dynamodb_cost")
	return utils.Get(dynamoDbCosts, SolverInputDynamodbReadCostDefault, "read_cost"), utils.Get(dynamoDbCosts, SolverInputDynamodbWriteCostDefault, "write_cost")
}

func (d *DataCenterLoader) GetEcrStorageCost(regionName string) float64 {
	return utils.Get(d.DataCenterData, SolverInputEcrMonthlyStorageCostDefault, regionName, "ecr_cost", "storage_cost")
}

func (d *DataCenterLoader) GetComputeCost(regionName string, architecture string) float64 {
	return utils.Get(d.DataCenterData, SolverInputComputeCostDefault, regionName, "execution_cost", "compute_cost", architecture)
}

func (d *DataCenterLoader) GetInvocationCost(regionName string, architecture string) float64 {
	return utils.Get(d.DataCenterData, SolverInputInvocationCostDefault, regionName, "execution_cost", "invocation_cost", architecture)
}

func (d *DataCenterLoader) GetTransmissionCost(regionName string, intraProviderTransfer bool) float64 {
	var transferType string
	if intraProviderTransfer {
		transferType = "provider_data_transfer"
	} else {
		transferType = "global_data_transfer"
	}
	return utils.Get(d.DataCenterData, SolverInputTransmissionCostDefault, regionName, "transmission_cost", transferType)
}
